File: ios/Sources/GutenbergKit/Sources/Model/EditorAssetBundle.swift
import CryptoKit
import Foundation
import SwiftSoup

/// A cached collection of editor assets downloaded from a remote manifest.
///
/// An `EditorAssetBundle` represents an on-disk cache of JavaScript and CSS assets
/// required by WordPress plugins and themes. The bundle is created by downloading
/// all assets specified in a server-provided manifest and storing them locally.
///
/// Bundles are identified by their manifest checksum, ensuring that different
/// versions of plugin/theme assets are stored separately. The `downloadDate`
/// property allows the system to prefer newer bundles over older ones.
///
/// Assets are accessed via URL lookup - the bundle maintains a mapping from
/// original remote URLs to local file paths.
public struct EditorAssetBundle: Sendable, Equatable, Hashable {


    /// The EditorRepresentation has the exact same format as `RemoteEditorAssetManifest.RawManifest` – what we're passing to Gutenberg
    /// looks exactly like what it'd get if it called `/wpcom/v2/editor-assets` directly.
    ///
    /// The difference is that we've rewritten all of the URLs to reference local files with our custom URL scheme so they can be provided from the on-disk cache.
    typealias EditorRepresentation = RemoteEditorAssetManifest.RawManifest

    /// Errors that can occur when working with asset bundles.
    enum Errors: Error, Equatable {
        /// The requested asset URL is not part of this bundle's manifest.
        case invalidRequest

        /// An asset with the same key already exists in the lookup table.
        case assetAlreadyExists(String)
    }

    /// The data structure stored on-disk
    struct RawAssetBundle: Codable {
        let manifest: LocalEditorAssetManifest
        let downloadDate: Date
    }

    /// The bundle's unique identifier, derived from its manifest checksum.
    ///
    /// Two bundles with the same ID have identical manifests and _should_ contain identical assets. This may not be
    /// true if a site is under development, so asset bundles should have some mechanism for being re-downloaded entirely.
    public var id: String {
        manifest.checksum
    }

    /// The manifest that defines which assets belong to this bundle.
    let manifest: LocalEditorAssetManifest

    /// The date this bundle was created by downloading the manifest contents.
    ///
    /// Used to determine which bundle is most recent when multiple bundles exist.
    let downloadDate: Date

    /// The number of assets stored in this bundle.
    public var assetCount: Int {
        manifest.assetUrls.count
    }

    let bundleRoot: URL

    init(raw: RawAssetBundle, bundleRoot: URL) {
        self.manifest = raw.manifest
        self.downloadDate = raw.downloadDate
        self.bundleRoot = bundleRoot
    }

    init(manifest: LocalEditorAssetManifest, downloadDate: Date = Date(), bundleRoot: URL) throws {
        self.manifest = manifest
        self.downloadDate = downloadDate
        self.bundleRoot = bundleRoot
    }

    /// Loads a bundle from a JSON file on disk.
    ///
    /// - Parameter url: The file URL of the bundle's `manifest.json`.
    /// - Throws: An error if the file cannot be read or decoded, or if required files are missing.
    init(url: URL) throws {
        let bundleRoot = url.deletingLastPathComponent()

        // Validate that editor-representation.json exists (required for the bundle to be usable)
        let editorRepPath = bundleRoot.appending(path: "editor-representation.json")
        guard FileManager.default.fileExists(atPath: editorRepPath.path) else {
            throw CocoaError(.fileNoSuchFile, userInfo: [NSFilePathErrorKey: editorRepPath.path])
        }

        self = try EditorAssetBundle(data: Data(contentsOf: url), bundleRoot: bundleRoot)
    }

    init(data: Data, bundleRoot: URL) throws {
        let rawBundle = try JSONDecoder().decode(RawAssetBundle.self, from: data)
        self = EditorAssetBundle(
            raw: rawBundle,
            bundleRoot: bundleRoot
        )
    }

    /// Checks whether the given asset URL resolves to a valid path within the bundle root.
    ///
    /// Use this method to validate URLs before calling `assetDataPath(for:)` or `hasAssetData(for:)`
    /// to avoid precondition failures for paths that escape the bundle root (e.g., plugin assets
    /// referenced in CSS that weren't downloaded into the bundle).
    ///
    /// - Parameter url: The asset URL to validate.
    /// - Returns: `true` if the URL resolves to a path within the bundle root, `false` otherwise.
    public func isValidAssetPath(for url: URL) -> Bool {
        let path = url.path(percentEncoded: false)
        let bundlePath = self.bundleRoot.appending(rawPath: path).standardizedFileURL
        let bundleRootPath = self.bundleRoot.standardizedFileURL.path
        return bundlePath.path.hasPrefix(bundleRootPath + "/") || bundlePath.path == bundleRootPath
    }

    /// Checks whether this bundle contains cached data for the given asset URL.
    ///
    /// - Parameter url: The original remote URL of the asset.
    /// - Returns: `true` if the asset is cached in this bundle, `false` otherwise.
    /// - Precondition: The URL path must not escape the bundle root directory.
    ///   Use `isValidAssetPath(for:)` to check validity before calling this method.
    public func hasAssetData(for url: URL) -> Bool {
        FileManager.default.fileExists(at: self.assetDataPath(for: url))
    }

    /// Returns the local file path for a cached asset.
    ///
    /// - Parameter url: The original remote URL of the asset.
    /// - Returns: The local file URL where the asset is stored.
    /// - Precondition: The URL path must not escape the bundle root directory.
    public func assetDataPath(for url: URL) -> URL {
        let path = url.path(percentEncoded: false)
        let bundlePath = self.bundleRoot.appending(rawPath: path).standardizedFileURL
        let bundleRootPath = self.bundleRoot.standardizedFileURL.path

        precondition(
            bundlePath.path.hasPrefix(bundleRootPath + "/") || bundlePath.path == bundleRootPath,
            "Asset path escapes bundle root: \(path)"
        )

        return bundlePath
    }

    /// Reads and returns the cached data for an asset.
    ///
    /// - Parameter url: The original remote URL of the asset.
    /// - Returns: The asset's file contents.
    /// - Throws: `Errors.invalidRequest` if the asset is not in this bundle,
    ///   or a file system error if the file cannot be read.
    public func assetData(for url: URL) throws -> Data {
        let fileURL = assetDataPath(for: url)
        return try Data(contentsOf: fileURL)
    }

    /// Reads the editor representation as a strongly-typed struct.
    ///
    /// The editor representation contains the rewritten script and style tags
    /// with URLs pointing to the local cache via the custom URL scheme.
    ///
    /// - Throws: An error if the file doesn't exist or cannot be decoded.
    func getEditorRepresentation() throws -> EditorRepresentation {
        let path = self.bundleRoot.appending(path: "editor-representation.json")
        let data = try Data(contentsOf: path)
        return try JSONDecoder().decode(EditorRepresentation.self, from: data)
    }

    /// Reads the editor representation as a JSON-serializable dictionary.
    ///
    /// Use this overload when you need to pass the representation to JavaScript.
    ///
    /// - Throws: An error if the file doesn't exist or cannot be parsed.
    func getEditorRepresentation() throws -> Any {
        let path = self.bundleRoot.appending(path: "editor-representation.json")
        let data = try Data(contentsOf: path)
        return try JSONSerialization.jsonObject(with: data)
    }

    /// Saves the editor representation to disk.
    ///
    /// - Parameter representation: The processed script/style tags with rewritten URLs.
    /// - Throws: An error if encoding or writing fails.
    func setEditorRepresentation(_ representation: EditorRepresentation) throws {
        let path = self.bundleRoot.appending(path: "editor-representation.json")
        try JSONEncoder().encode(representation).write(to: path, options: .atomic)
    }

    /// Returns the bundle's manifest as JSON data for storage.
    func dataRepresentation() throws -> Data {
        try JSONEncoder().encode(RawAssetBundle(
            manifest: self.manifest,
            downloadDate: self.downloadDate
        ))
    }

    /// Writes the bundle's JSON representation to disk.
    ///
    /// - Parameter path: The file URL where the bundle should be saved.
    /// - Throws: An error if encoding fails or the file cannot be written.
    func writeManifest(to path: URL? = nil, editorRepresentation: EditorRepresentation? = nil) throws {
        try FileManager.default.createDirectory(at: self.bundleRoot, withIntermediateDirectories: true)
        let destination = path ?? self.bundleRoot.appendingPathComponent("manifest.json")
        try self.dataRepresentation().write(to: destination, options: .atomic)

        if let editorRepresentation {
            try setEditorRepresentation(editorRepresentation)
        }
    }

    /// Copies the bundle to the given directoy.
    ///
    /// APFS makes this instant and zero-cost.
    func copy(to destination: URL) throws -> EditorAssetBundle {

        // Don't bother persisting an empty bundle
        guard self != .empty else {
            return .empty
        }

        if FileManager.default.directoryExists(at: destination) {
            try FileManager.default.removeItem(at: destination)
        }

        let destinationParent = destination.deletingLastPathComponent()
        try FileManager.default.createDirectory(at: destinationParent, withIntermediateDirectories: true)
        try FileManager.default.copyItem(at: self.bundleRoot, to: destination)

        return try EditorAssetBundle(url: destination.appending(path: "manifest.json"))
    }

    static let empty: EditorAssetBundle = EditorAssetBundle(
        raw: RawAssetBundle(
            manifest: .empty,
            downloadDate: Date()
        ),
        bundleRoot: URL.temporaryDirectory
    )
}

File: android/Gutenberg/src/main/java/org/wordpress/gutenberg/model/EditorAssetBundle.kt
package org.wordpress.gutenberg.model

import android.util.Log
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.File
import java.util.Date

private const val TAG = "EditorAssetBundle"

/**
 * A collection of editor assets downloaded from a WordPress site.
 *
 * The asset bundle contains both the manifest (describing what assets are needed)
 * and serves as the root directory for cached asset files. Assets are stored
 * in a directory structure mirroring their original URL paths.
 *
 * ## Lifecycle
 *
 * 1. A bundle is created with a manifest from the server
 * 2. Assets are downloaded and stored within the bundle's root directory
 * 3. The bundle is persisted to disk with its metadata
 * 4. On subsequent launches, the bundle can be loaded from disk and validated
 *
 * ## Identification
 *
 * Each bundle has an `id` derived from its manifest's checksum. This ensures that
 * when a site's assets change, a new bundle will be created with a different ID,
 * allowing old bundles to be cleaned up.
 */
data class EditorAssetBundle(
    /** The parsed manifest describing the assets in this bundle. */
    val manifest: LocalEditorAssetManifest,
    /** When this bundle was downloaded from the server. */
    val downloadDate: Date,
    /** The root directory where this bundle's assets are stored. */
    val bundleRoot: File
) {
    /**
     * A unique identifier for this bundle based on its manifest content.
     *
     * Returns "empty" for bundles with empty manifests, otherwise returns
     * the manifest's SHA-256 checksum.
     */
    val id: String
        get() = manifest.checksum

    /**
     * The total number of assets (scripts + styles) that need to be cached.
     */
    val assetCount: Int
        get() = manifest.assetUrls.size

    /**
     * A raw bundle structure for JSON serialization.
     */
    @Serializable
    data class RawAssetBundle(
        val manifest: LocalEditorAssetManifest,
        val downloadDate: Long // Unix timestamp in milliseconds
    ) {
        constructor(manifest: LocalEditorAssetManifest, downloadDate: Date) : this(
            manifest = manifest,
            downloadDate = downloadDate.time
        )

        fun toEditorAssetBundle(bundleRoot: File): EditorAssetBundle {
            return EditorAssetBundle(
                manifest = manifest,
                downloadDate = Date(downloadDate),
                bundleRoot = bundleRoot
            )
        }
    }

    companion object {
        private val json = Json { ignoreUnknownKeys = true }
        private const val MANIFEST_FILENAME = "manifest.json"
        private const val EDITOR_REPRESENTATION_FILENAME = "editor-representation.json"

        /**
         * An empty bundle for sites that don't support the editor-assets endpoint.
         */
        val empty = EditorAssetBundle(
            manifest = LocalEditorAssetManifest.empty,
            downloadDate = Date(),
            bundleRoot = File("/tmp/empty-bundle")
        )

        /**
         * Loads a bundle from a manifest file URL.
         *
         * @param file The file containing the serialized bundle manifest.
         * @throws Exception if the file cannot be read or parsed.
         */
        fun fromFile(file: File): EditorAssetBundle {
            val data = file.readText()
            val rawBundle = json.decodeFromString<RawAssetBundle>(data)
            return rawBundle.toEditorAssetBundle(file.parentFile ?: file)
        }
    }

    /**
     * Creates a bundle with the current time as the download date.
     */
    constructor(manifest: LocalEditorAssetManifest, bundleRoot: File) : this(
        manifest = manifest,
        downloadDate = Date(),
        bundleRoot = bundleRoot
    )

    /**
     * Checks if the asset data exists for a given URL.
     *
     * @param url The original asset URL.
     * @return `true` if the asset has been cached locally.
     */
    fun hasAssetData(url: String): Boolean {
        if(this == empty) {
            return false
        }

        return assetDataPath(url).exists()
    }

    /**
     * Returns the local file path where an asset should be stored.
     *
     * The path is constructed by appending the URL's path component to
     * the bundle's root directory.
     *
     * Note: This method must not be called on [EditorAssetBundle.empty].
     * In debug builds, an assertion will catch this misuse.
     *
     * @param url The original asset URL.
     * @return The file where the asset is (or should be) stored.
     * @throws IllegalArgumentException if the URL path would escape the bundle root.
     */
    fun assetDataPath(url: String): File {
        assert(this != empty) { "Cannot get asset path from empty bundle" }

        val urlPath = java.net.URL(url).path
        val resolvedFile = File(bundleRoot, urlPath).canonicalFile
        val bundleRootCanonical = bundleRoot.canonicalFile

        require(
            resolvedFile.path.startsWith(bundleRootCanonical.path + File.separator) ||
                resolvedFile.path == bundleRootCanonical.path
        ) {
            "Asset path escapes bundle root: $urlPath"
        }

        return resolvedFile
    }

    /**
     * Reads the cached asset data for a given URL.
     *
     * @param url The original asset URL.
     * @return The asset's binary data.
     * @throws Exception if the asset has not been cached.
     */
    fun assetData(url: String): ByteArray {
        return assetDataPath(url).readBytes()
    }

    /**
     * Stores the editor representation for later retrieval.
     *
     * The editor representation is the processed manifest with rewritten URLs
     * ready for injection into the WebView.
     *
     * @param representation The processed manifest to store.
     */
    fun setEditorRepresentation(representation: RemoteEditorAssetManifest.RawManifest) {
        val file = File(bundleRoot, EDITOR_REPRESENTATION_FILENAME)
        val data = json.encodeToString(representation)
        file.writeText(data)
        Log.d(TAG, "Wrote editor representation: file=${file.absolutePath}, size=${data.length} bytes")
    }

    /**
     * Retrieves the stored editor representation.
     *
     * @return The stored representation.
     * @throws Exception if no representation has been stored.
     */
    fun getEditorRepresentation(): RemoteEditorAssetManifest.RawManifest {
        val file = File(bundleRoot, EDITOR_REPRESENTATION_FILENAME)
        val data = file.readText()
        return json.decodeFromString<RemoteEditorAssetManifest.RawManifest>(data)
    }

    /**
     * Retrieves the stored editor representation as a map for JSON serialization.
     *
     * @return A map representation suitable for JSON injection.
     * @throws Exception if no representation has been stored.
     */
    fun getEditorRepresentationAsMap(): Map<String, Any> {
        val rep = getEditorRepresentation()
        return mapOf(
            "scripts" to rep.scripts,
            "styles" to rep.styles,
            "allowed_block_types" to rep.allowedBlockTypes
        )
    }

    /**
     * Saves the bundle's metadata to disk.
     *
     * Call this after downloading all assets to persist the bundle for future sessions.
     */
    fun save() {
        val rawBundle = RawAssetBundle(manifest = manifest, downloadDate = downloadDate)
        val file = File(bundleRoot, MANIFEST_FILENAME)
        val data = json.encodeToString(rawBundle)
        file.writeText(data)
        Log.d(TAG, "Wrote asset bundle manifest: file=${file.absolutePath}, size=${data.length} bytes")
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is EditorAssetBundle) return false
        return manifest == other.manifest && downloadDate == other.downloadDate
    }

    override fun hashCode(): Int {
        var result = manifest.hashCode()
        result = 31 * result + downloadDate.hashCode()
        return result
    }
}
